File: java/exercises/02-intermediate/ex037_generic_box/BoxTest.java
package fewolearning.exercises.intermediate.ex037_generic_box;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class BoxTest {

    @Test
    void isEmptyIsTrueForANewBox() {
        Box<String> box = new Box<>();

        assertTrue(box.isEmpty());
    }

    @Test
    void setStoresAValueAndClearsTheEmptyFlag() {
        Box<String> box = new Box<>();

        box.set("hello");

        assertFalse(box.isEmpty());
    }

    @Test
    void getReturnsTheStoredValue() {
        Box<Integer> box = new Box<>();
        box.set(42);

        assertEquals(42, box.get());
    }

    @Test
    void getThrowsWhenTheBoxIsEmpty() {
        Box<String> box = new Box<>();

        assertThrows(IllegalStateException.class, box::get);
    }
}

File: kotlin/exercises/01-beginner/ex025_generic_box/BoxTest.kt
package fewolearning.exercises.beginner.ex025_generic_box

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull

class BoxTest {

    @Test
    fun getReturnsNullForAFreshlyConstructedEmptyBox() {
        val box = Box<String>()

        assertNull(box.get())
    }

    @Test
    fun setThenGetReturnsTheStoredValue() {
        val box = Box<String>()

        box.set("hello")

        assertEquals("hello", box.get())
    }

    @Test
    fun getReturnsTheInitialConstructorValueWhenProvided() {
        val box = Box("start")

        assertEquals("start", box.get())
    }

    @Test
    fun setOverwritesAPreviouslyStoredValue() {
        val box = Box(1)

        box.set(2)

        assertEquals(2, box.get())
    }
}
